include/circt/Dialect/SMT/SMTOps.td

```diff
@@ -16,8 +16,152 @@ include "mlir/IR/EnumAttr.td"
 include "mlir/IR/OpAsmInterface.td"
 include "mlir/Interfaces/InferTypeOpInterface.td"
 include "mlir/Interfaces/SideEffectInterfaces.td"
+include "mlir/Interfaces/ControlFlowInterfaces.td"
 
 class SMTOp<string mnemonic, list<Trait> traits = []> :
   Op<SMTDialect, mnemonic, traits>;
 
+def DeclareConstOp : SMTOp<"declare_const", [
+  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>
+]> {
+  let summary = "declare a symbolic value of a given sort";
+  let description = [{
+    This operation declares a symbolic value just as the `declare-const`
+    statement in SMT-LIB 2.6. The result type determines the SMT sort of the
+    symbolic value.
+    Unlike the SMT-LIB version, this operation allows for a `fresh` attribute to
+    indicate that a new unique value should be generated. The provided string
+    will then be used as a prefix for the newly generated identifier. If this
+    attribute is not present and there are two such operations
+    `%a = smt.declare_const "a"` and `%b = smt.declare_const "a"` then `%a` and
+    `%b` are equivalent, i.e., they refer to the same symbolic value.
+
+    Note that this operation cannot be unconditionally marked as Pure since two
+    operations with the `fresh` attribute and the same identifier string could
+    then be CSEd, leading to incorrect behavior.
```

Comment:
Does this need to have two separate SSA values that refer to the same declared constant ever arise? Could it make sense to always create a fresh constant, and then mark this operation pure? I could imagine that the aliasing/reuse of same name constants is more of a SMT-LIB writing convenience, but if you're lowering to that from an SSA representation, you might as well enforce a single definition rule.

Reply:
Good point. I agree that we should always create a fresh symbolic value. This non-fresh behavior was the initial implementation modeled too closely to an AST.
We cannot mark it 'pure' though since two symbolic values with the same prefix string will represent separate symbolic values and them being pure would allow CSE to merge them, i.e., change the behavior.

Reply:
I now declare the `declare_const` operation's result as a memory allocation, so it is DCEd but not CSEd.